crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -1151,6 +1151,82 @@ impl<'w> EntityWorldMut<'w> {
         self
     }
 
+    /// Removes a dynamic [`Component`] from the entity.
+    ///
+    /// You should prefer to use the typed API [`EntityWorldMut::remove`] where possible.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the provided [`ComponentId`] does not exist in the [`World`].
+    pub fn remove_by_id(&mut self, component_id: ComponentId) -> &mut Self {
+        let archetypes = &mut self.world.archetypes;
+        let storages = &mut self.world.storages;
+        let components = &mut self.world.components;
+        let entities = &mut self.world.entities;
+        let removed_components = &mut self.world.removed_components;
+
+        let bundle_id = self
+            .world
+            .bundles
+            .init_component_info(components, component_id);
+
+        let Some(bundle_info) = self.world.bundles.get(bundle_id) else {
+            // if no bundle info, certainly no components to remove
+            return self;
+        };
+
+        let old_location = self.location;
+        // SAFETY: `archetype_id` exists because it is referenced in the old `EntityLocation` which is valid,
+        // components exist in `bundle_info` because `Bundles::init_component_info` would have panicked otherwise
+        let new_archetype_id = unsafe {
+            remove_bundle_from_archetype(
+                archetypes,
+                storages,
+                components,
+                old_location.archetype_id,
+                bundle_info,
+                true,
+            )
+            .expect("intersections should always return a result")
+        };
+
+        if new_archetype_id == old_location.archetype_id {
+            return self;
+        }
+
+        let old_archetype = &mut archetypes[old_location.archetype_id];
+        let entity = self.entity;
+        if old_archetype.contains(component_id) {
```

Comment:
@SkiFire13 from the last pr: "If the archetype changed then this contains should always be true"

Just to clarify should I remove the if statement?

Reply:
Yes, the comment still holds.

The code was probably adapted from `EntityMut::remove_bundle` where there may be multiple sparse set components in the bundle, so a change in the archetype doesn't necessarily mean the component was actually removed, as it may not have been present before but the archetype could still have changed due to removing a different component. In this case however there's only one component, so if the archetype changed then the component must have been removed.